### PR TITLE
WIP: Enable existing options for stand-alone instrument command

### PR DESCRIPTION
**Pull request by AndrewFinlay**
Make include, exclude, cwd and es-modules options available for the stand-alone instrument command to make the behaviour more closely resemble that of the standard nyc command. 

This is aimed at fixing #596 and #890. 

This change allows access to these config options making stand alone instrumentation more flexible.

These options I've added here are taken from config-util.js, with a few changes made to the descriptions.  Also, I've removed the global flag in keeping with the other instrument options but I'm not entirely sure of the effect of this so it may need to change.

### CODE REVIEW

**Comment by JaKXz** [comment]
🙌

**Comment by AndrewFinlay** [comment]
@coreyfarrell - It seems that the PR has gone stale with regards to the default es-modules option, will fix that.  Regarding having the options declared only once, it's a good idea but that sounds like a major redesign of the current nyc configuration system and is well beyond the scope of this PR.

@JaKXz I've had another dig though the issue log and couldn't find much more history that I hadn't already referenced.  I did find a discussion between @bcoe and @gotwarlost that providing coverage tools for client-side coverage was a lower priority in terms of the initial nyc release, #293 and #309, but I don't think it's really relevant now.  Off the top of your head is there anything in particular that you know is relevant to this PR?

Regarding the robustness of the instrument command, I haven't dug too deeply but it does seem to have a slightly different implementation to that used in the general nyc command.  Not sure about testing for this, I'll have another look at bcoe's earlier work

**Comment by AndrewFinlay** [comment]
Hi all, this PR/issue seems to have lost momentum and I was hoping to get it moving again.  What do I need to do to get this over the line?

**Comment by bcoe** [comment]
@AndrewFinlay @JaKXz will defer to @coreyfarrell's review, a unit test exercising the new behavior would be nice.

very much appreciate the contribution, as an aside feel free to join the community [here](http://devtoolscommunity.herokuapp.com/), if we're ever taking a long time to get back ahold of you.

**Comment by AndrewFinlay** [comment]
Bump, stale bot

**Comment by JaKXz** [comment]
Have the comments on this PR been addressed? The change makes sense to me at a high level

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ All notable changes to this project will be documented in this file. See [standa
 
 ### Bug Fixes
 
-* add flag to allow control of intrumenter esModules option, default to looser parsing ([#863](https://github.com/istanbuljs/nyc/issues/863)) ([6b6cd5e](https://github.com/istanbuljs/nyc/commit/6b6cd5e))
+* add flag to allow control of instrumenter esModules option, default to looser parsing ([#863](https://github.com/istanbuljs/nyc/issues/863)) ([6b6cd5e](https://github.com/istanbuljs/nyc/commit/6b6cd5e))
```
Comment (coreyfarrell) [inline CHANGELOG.md]: @bcoe does this create a problem with the utility which generates the CHANGELOG.md?

Really the problem is that the typo is in the commit log, I'm not sure this can be fixed?

File: test/nyc-bin.js [hunk 2]
```diff
@@ -561,10 +563,130 @@ describe('the nyc cli', function () {
...
+
+          var stderr = ''
+          proc.stderr.on('data', function (chunk) {
+            stderr += chunk
+          })
+
+          proc.on('close', function (code) {
+            code.should.equal(1)
+            stdoutShouldEqual(stderr, `
+        Failed to instrument ./not-strict.js`)
```
Comment (coreyfarrell) [inline test/nyc-bin.js]: nit: Could you increase the indent so that the line with `Failed to instrument` is indented 2 spaces more than `stdoutShouldEqual`?

File: CHANGELOG.md (hunk 1 above)
Comment (AndrewFinlay) [reply CHANGELOG.md]: Sorry, didn't realize that would be a problem and think to look if it was generated, I was on typo autopilot there.  
If it's just going to disappear next time the tool runs then we'll remove it

Comment (bcoe) [reply CHANGELOG.md]: the changelog generation looks at the header, doesn't re-generate the whole document, so the edit should be fine.

File: index.js [hunk 3]
```diff
@@ -200,6 +200,11 @@ NYC.prototype.instrumentAllFiles = function (input, output, cb) {
     var ext
     var transform
     var inFile = path.resolve(inputDir, filename)
+
+    if (!_this.exclude.shouldInstrument(filename)) {
```
Comment (bcoe) [inline index.js]: refresh me if I'm wrong about how source-maps behave, but is it possible that a filename on disk would be excluded by our exclude rule, but has a filename the source-map maps to that isn't excluded?

we should look at the tests around excludeAfterInstrument behavior, which addresses some of these edge-cases, and make sure we continue to support these edge-cases.

Comment (coreyfarrell) [reply index.js]: Maybe this should be a separate change?  I'm not sure how it's related to the main purpose of this PR 'enable existing options for stand-alone instrument command'.

File: test/nyc-bin.js (hunk 2 above)
Comment (AndrewFinlay) [reply test/nyc-bin.js]: Didn't realise that was running through a strip indent, will fix in the next commit.

**Comment by AndrewFinlay** [comment]
I've added the clean option to this PR, the only issue is that it's a breaking change, changing the current default behaviour from leaving the existing output dir untouched to removing any existing output and starting again.  I figure having 'clean' default to true is the expected behaviour, but that's not really my decision to make so feedback on that would be appreciated.

I also implemented this in a pretty simple way, I figured I should check to see if you want this to use the clean command on the Nyc object.  If so there's a few hoops to jump through to get that to work.

File: lib/commands/instrument.js [hunk 4]
```diff
@@ -51,14 +56,37 @@ exports.builder = function (yargs) {
...
+      default: cwd
+    })
+    .option('es-modules', {
+      default: true,
+      type: 'boolean',
+      describe: 'tell the instrumenter to treat files as ES Modules'
+    })
+    .option('clean', {
+      describe: 'should the output folder be cleaned before instrumenting files?',
+      default: true,
```
[the remark this answers is not in the document]
Comment (AndrewFinlay) [reply lib/commands/instrument.js]: Hi Corey, I've tried to write this a few times but it was getting a little wordy. So quite simply, I don't feel there's much point in releasing this without the exclude, include and includeNegated changes.  So if it needs to wait for a major release then that's what should happen.  That way it can be part of a comprehensive effort to address missing functionality in the instrument command and be a PR that you're satisfied doesn't have any issues.

File: index.js (hunk 3 above)
Comment (AndrewFinlay) [reply index.js]: @coreyfarrell This change is pretty much the core of enabling include, exclude, excludeNegated functionality.  An assessment of whether the file should be instrumented is conducted in the function `_this.exclude.shouldInstrument(filename)`.  If it shouldn't be instrumented we return, if it should we continue and instrument the file.

@bcoe I've been looking over anything source-maps related and I'm still working on getting a good understanding of how everything hangs together.  So far I can see that the 'produce-source-maps' adds a source map while we're instrumenting, which shouldn't be an issue and seems unrelated to the 'source-map' option. 
I'm a little unsure of what 'source-map' should do in the instrument command.  My understanding is that it's mainly used to work with pre-instrumented files with existing source maps but I don't know why you'd be running the instrument command if you'd already instrumented the code elsewhere.  Anyway, my plan is to keep working on this until I'm confident it's good, any pointers, gotchas, or ideas for tests on this stuff would be appreciated.  Thanks

File: lib/commands/instrument.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bcoe) [reply lib/commands/instrument.js]: If we do wait for the next major release, might be worth revisiting some of the exclude logic in general; it's a real pain in the butt when combined with source-maps -- I'd argue it might be worth expanding on the test suite to cover all these weird edge-cases, so that it's easier to add new features like this in the future, without bumping into foot guns (i.e., me and @coreyfarrell not being sure how safe the change is).

**Comment by AndrewFinlay** [comment]
Hi all, I'd like to close this PR and break it up into more manageable pieces, any objections?
